File: Braille_Algorithm_Development/CellParser/GradeOneLUTs.py
prefixes = {
    8: "ampersand_angleBracket_currency_accent_tilde_mathPower_@_scriptIndicator_dagger_prefix",
    16: "asterisk_roundBracket_basicMathSign_ditto_dash_prefix",
    24: "genderSign_doubleQuotation_IP_accent_paragraphSection_degree_boldIndicator_ligaturedIndicator_prefix",
    32: "capital_singleQuotation_prefix",
    40: "squareBracket_percentage_greekLetter_underscore_italicIndicator_generalFractionline_prefix",
    43: "shapeIndicator_prefix",
    51: "arrow_Prefix",
    54: "feet_inches_prefix",
    56: "curlyBracket_bulletPoint_hashTag_slash_guillemet_underlineIndicator_prefix",
    60: "number_musicSymbol_prefix",
}

prefix_8 = {
    47: u'\u0026',  # "ampersand"
    35: u'\u3008',  # "openingAngleBracket"
    28: u'\u3009',  # "closingAngleBracket"
    9: u'\u00A2',  # "cent"
    14: u'\u0024',  # "dollar"
    17: u'\u20AC',  # "euro"
    7: u'\u00A3',  # "pound"
    61: u'\u00A5',  # "yen"

    # Accented letters
    # if letter is capital capital indicator comes before accent indicator which in turn comes before the letter
    44: "breve",
    36: "macron",

    20: u'\u007E',  # "tilde" - used for approximately
    34: u'\u005E',  # "mathPower"8, 1
    1: u'\u0040',  # "@"
    32: "daggerIndicator",

    6: "scriptSymbol",
    2: "scriptWord",
    54: "scriptPassage",
    4: "scriptTerminator",
}

prefix_8_32 = {
    57: u'\u2020',  # "dagger"
    59: u'\u2021',  # "doubleDagger"
}

prefix_16 = {
    20: u'\u002A',  # "asterisk"
    35: u'\u0028',  # "openingRoundBracket"
    28: u'\u0029',  # "closingRoundBracket"
    22: u'\u002B',  # "additionSign"
    38: u'\u00D7',  # "multiplicationSign"
    12: u'\u00F7',  # "divisionSign"
    54: u'\u003D',  # "equalsSign"
    2: u'\u3003',  # "ditto"
    32: "emDashIndicator",
    36: u'\u2212',  # "enDash_minusSign"
}

prefix_16_32 = {
    36: u'\u2014',  # "emDash"
}

prefix_24 = {
    45: u'\u2640',  # "femaleGenderSign",
    61: u'\u2642',  # "maleGenderSign",
    38: u'\u201C',  # "openingDoubleQuotation"
    52: u'\u201D',  # "closingDoubleQuotation"
    9: u'\u00A9',  # "copyright"
    23: u'\u00AE',  # "registered"
    30: u'\u2122',  # "trademark"

    # Accented letters
    # if letter is capital capital indicator comes before accent indicator which in turn comes before the letter
    12: "acute",
    44: "caron",
    47: "cedilla",
    41: "circumflex",
    18: "diaeresis",
    33: "grave",
    43: "ring",
    59: "tilde",

    15: u'\u2761',  # "paragraph"
    14: u'\u00A7',  # "section"
    26: u'\u00B0',  # "degree"

    6: "boldSymbol",
    2: "boldWord",
    54: "boldPassage",
    4: "boldTerminator",

    22: "ligatureIndicator",
}

# could have signified a capital letter if subsequent cell is not one of these
prefix_32 = {
    38: u'\u2018',  # "openingSingleQuotation",
    52: u'\u2019',  # "closingSingleQuotation",
    32: "capitalWord",
    4: "capitalTerminator"
}

prefix_32_32 = {
    32: "capitalPassage"
}

prefix_40 = {
    35: u'\u005B',  # "openingSquareBracket"
    28: u'\u005D',  # "closingSquareBracket"
    52: u'\u0025',  # "percentage"
    36: u'\u005F',  # "underscore"
    12: u'\u2044',  # "generalFractionLine"

    6: "italicSymbol",
    2: "italicWord",
    54: "italicPassage",
    4: "italicTerminator",

    # greek
    # comes before every greek letter
    # if letter is capital - capital indicator (32) comes before each indicator
    1: u'\u03B1',  # alpha
    3: u'\u03B2',  # "beta"
    27: u'\u03B3',  # "gamma"
    25: u'\u03B4',  # "delta"
    17: u'\u03B5',  # "epsilon"
    53: u'\u03B6',  # "zeta"
    49: u'\u03B7',  # "eta"
    57: u'\u03B8',  # "theta"
    10: u'\u03B9',  # "iota"
    5: u'\u03BA',  # "kappa"
    7: u'\u03BB',  # "lambda"
    13: u'\u03BC',  # "mu"
    29: u'\u03BD',  # "nu"
    45: u'\u03BE',  # "xi"
    21: u'\u03BF',  # "omicron"
    15: u'\u03C0',  # "pi"
    23: u'\u03C1',  # "rho"
    14: u'\u03C3',  # "sigma"
    30: u'\u03C4',  # "tau"
    37: u'\u03C5',  # "upsilon"
    11: u'\u03C6',  # "phi"
    47: u'\u03C7',  # "chi"
    61: u'\u03C8',  # "psi"
    58: u'\u03C9',  # "omega"
}

prefix_43 = {
    63: u'\u25CF',  # "circle"
    60: "squareTriangleIndicator"
}

prefix_43_60 = {
    25: u'\u25A0',  # "square"
    9: u'\u25B2',  # "triangle"
}

# Arrows
prefix_51 = {
    21: u'\u2192',  # "rightArrow"
    42: u'\u2190',  # "leftArrow"
    41: u'\u2193',  # "downArrow"
    44: u'\u2191',  # "upArrow"
    14: u'\u2197',  # "upRightArrow"
    35: u'\u2198',  # "downRightArrow"
    28: u'\u2199',  # "downLeftArrow"
    49: u'\u2196',  # "upLeftArrow"
}

# not here - feet
prefix_54 = {
    54: u'\u2033'  # " (inches)
}

prefix_56 = {
    35: u'\u007B',  # "openingCurlyBracket"
    28: u'\u007D',  # "closingCurlyBracket"
    50: u'\u2022',  # "bulletPoint"
    57: u'\u0023',  # "hashTag"
    33: u'\u2216',  # "backwardSlash"
    12: u'\u2215',  # "forwardSlash"
    38: u'\u00AB',  # "openingGuillemet"
    52: u'\u00BB',  # "closingGuillement"

    6: "underlineSymbol",
    2: "underlineWord",
    54: "underlinePassage",
    4: "underlineTerminator",
}

# UEB Music
prefix_60 = {
    35: u'\u266D',  # "flat"
    33: u'\u266E',  # "natural"
    41: u'\u266F',  # "sharp"
}

numberMode = {
    48: "grade1Indicator",  # used if number is followed by a lowercase letter from "a-j" (even if there is a "." or "," after number
    16: "numericSpace"
}

letters = {
    1: u'\u0061',  # a
    3: u'\u0062',  # b
    9: u'\u0063',  # c
    25: u'\u0064',  # d
    17: u'\u0065',  # e
    11: u'\u0066',  # f
    27: u'\u0067',  # g
    19: u'\u0068',  # h
    10: u'\u0069',  # i
    26: u'\u006A',  # j
    5: u'\u006B',  # k
    7: u'\u006C',  # l
    13: u'\u006D',  # m
    29: u'\u006E',  # n
    21: u'\u006F',  # o
    15: u'\u0070',  # p
    31: u'\u0071',  # q
    23: u'\u0072',  # r
    14: u'\u0073',  # s
    30: u'\u0074',  # t
    37: u'\u0075',  # u
    39: u'\u0076',  # v
    58: u'\u0077',  # w
    45: u'\u0078',  # x
    61: u'\u0079',  # y
    53: u'\u007A'  # z
}

numbers = {
    26: u'\u0030',
    1: u'\u0031',
    3: u'\u0032',
    9: u'\u0033',
    25: u'\u0034',
    17: u'\u0035',
    11: u'\u0036',
    27: u'\u0037',
    19: u'\u0038',
    10: u'\u0039'
}

basicPunctuation = {
    4: "apostrophe_capitalTerminator",  # capitalTerminator only required when following a capital prefix (word/passage)
    2: u'\u002C',  # comma
    36: u'\u002D',  # hyphen
    50: u'\u002E',  # full stop - 3 in a row is an elipsis
    22: u'\u0021',  # exclamation mark
    6: u'\u003B',  # semi-colon
    18: u'\u003A',  # colon
    38: "questionMark_openingNonSpecificQuotation",
    52: "'",  #
    41: u'\u221A',  # "openSquareRoot",
    44: "closeSquareRoot",
    34: "subscriptIndicator",
    20: "superscriptIndicator",
    12: u'\u2044',  # "simpleNumericFractionLine"
    55: "generalFractionOpenIndicator",
    62: "generalFractionCloseIndicator",
    49: "shapeTerminator"  # not required if shape is followed by space (only followed by letter, number or punctuation)
}

File: Braille-Assistant/app/src/main/java/org/example/brailleassistant/processors/GradeOneLUTs.java
package org.example.brailleassistant.processors;

import java.util.HashMap;

// Look up tables for grade 1 braille translation
public class GradeOneLUTs {

    final public int[] prefixes = {8, 16, 24, 32, 40, 43, 51, 54, 56, 60};

    public HashMap<Integer, String> prefix_8;

    public HashMap<Integer, String> prefix_8_32;

    public HashMap<Integer, String> prefix_16;

    public HashMap<Integer, String> prefix_16_32;

    public HashMap<Integer, String> prefix_24;

    public HashMap<Integer, String> prefix_32;

    public HashMap<Integer, String> prefix_32_32;

    public HashMap<Integer, String> prefix_40;

    public HashMap<Integer, String> prefix_43;

    public HashMap<Integer, String> prefix_43_60;

    public HashMap<Integer, String> prefix_51;

    public HashMap<Integer, String> prefix_54;

    public HashMap<Integer, String> prefix_56;

    public HashMap<Integer, String> prefix_60;

    public HashMap<Integer, String> numberMode;

    public HashMap<Integer, String> letters;

    public HashMap<Integer, String> numbers;

    public HashMap<Integer, String> punctuation;

    //    ACCENT INTEGER ENCODING
    //    201 - acute
    //    202 - caron
    //    203 - cedilla
    //    204 - circumflex
    //    205 - diaeresis
    //    206 - grave
    //    207 - ring
    //    208 - tilde
    //    209 - breve
    //    210 - macron
    //    211 - ligature
    //    212 - subscript
    //    213 - superscript

    public GradeOneLUTs() {

        prefix_8 = new HashMap<Integer, String>() {{

            put(47, "\u0026"); // ampersand
            put(35, "\u3008"); // opening angle bracket
            put(28, "\u3009"); // closing angle bracket
            put(9, "\u00A2"); // cent
            put(14, "\u0024"); // dollar
            put(17, "\u20AC"); // euro
            put(7, "\u00A3"); // pound
            put(61, "\u00A5"); // yen

            put(44, "209"); // breve
            put(36, "210"); // macron

            put(20, "\u007E"); // tilde
            put(34, "\u005E"); // math power
            put(1, "\u0040"); // @
            put(32, "DAGGER_INDICATOR");

            put(6, "<ss>"); // SCRIPT_SYMBOL
            put(2, "<sw>"); // SCRIPT_WORD
            put(54, "<s>"); // SCRIPT_PASSAGE
            put(4, "</s>"); // SCRIPT_TERMINATOR

        }};

        prefix_8_32 = new HashMap<Integer, String>() {{

            put(57, "\u2020"); // dagger
            put(59, "\u2021"); // double dagger

        }};

        prefix_16 = new HashMap<Integer, String>() {{

            put(20, "\u002A"); // asterisk
            put(35, "\u0028"); // opening round bracket
            put(28, "\u0029"); // closing round bracket
            put(22, "\u002B"); // addition sign
            put(38, "\u00D7"); // multiplication sign
            put(12, "\u00F7");  // division sign
            put(54, "\u003D"); // equals sign
            put(2, "\u3003"); // ditto
            put(32, "EM_DASH_INDICATOR");
            put(36, "\u2212"); // en dash or minus sign

        }};

        prefix_16_32 = new HashMap<Integer, String>() {{

            put(36, "\u2014"); // em dash

        }};

        prefix_24 = new HashMap<Integer, String>() {{

            put(45, "\u2640"); // female gender sign
            put(61, "\u2642"); // male gender sign
            put(38, "\u201C"); // opening double quotation
            put(52, "\u201D"); // closing double quotation
            put(9, "\u00A9"); // copyright
            put(23, "\u00AE"); // registered
            put(30, "\u2122"); // trademark

            // Accented letters
            //  if letter is capital capital indicator comes before accent indicator which in turn comes before the letter
            put(12, "201"); // acute
            put(44, "202"); // caron
            put(47, "203"); // cedilla
            put(41, "204"); // circumflex
            put(18, "205"); // diaeresis
            put(33, "206"); // grave
            put(43, "207"); // ring
            put(59, "208"); // tilde

            put(15, "\u2761"); // paragraph
            put(14, "\u00A7"); // section
            put(26, "\u00B0"); // degree

            put(6, "<bs>"); // BOLD_SYMBOL
            put(2, "<bw>"); // BOLD_WORD
            put(54, "<b>"); // BOLD_PASSAGE
            put(4, "</b>"); // BOLD_TERMINATOR

            put(22, "211"); // LIGATURE

        }};

        prefix_32 = new HashMap<Integer, String>() {{

            put(38, "\u2018"); // opening single quotation
            put(52, "\u2019"); // closing single quotation
            put(32, "<cw>"); // CAPITAL_WORD
            put(4, "</c>"); // CAPITAL_TERMINATOR

        }};

        prefix_32_32 = new HashMap<Integer, String>() {{

            put(32, "<c>"); // CAPITAL_PASSAGE

        }};

        prefix_40 = new HashMap<Integer, String>() {{

            put(35, "\u005B"); // opening square Bracket
            put(28, "\u005D"); // closing square bracket
            put(52, "\u0025"); // percentage
            put(36, "\u005F"); // underscore
            put(12, "\u2044"); // general fraction line

            put(6, "<is>"); // ITALIC_SYMBOL
            put(2, "<iw>"); // ITALIC_WORD
            put(54, "<i>"); // ITALIC_PASSAGE
            put(4, "</i>"); // ITALIC_TERMINATOR

            // greek letters
            // prefix comes before every greek letter
            // if letter is capital - capital indicator (32) comes before each greek prefix indicator
            put(1, "\u03B1"); // alpha
            put(3, "\u03B2"); // beta
            put(27, "\u03B3"); // gamma
            put(25, "\u03B4"); // delta
            put(17, "\u03B5"); // epsilon
            put(53, "\u03B6"); // zeta
            put(49, "\u03B7"); // eta
            put(57, "\u03B8"); // theta
            put(10, "\u03B9"); // iota
            put(5, "\u03BA"); // kappa
            put(7, "\u03BB");  // lambda
            put(13, "\u03BC"); // mu
            put(29, "\u03BD"); // nu
            put(45, "\u03BE"); // xi
            put(21, "\u03BF"); // omicron
            put(15, "\u03C0"); // pi
            put(23, "\u03C1"); // rho
            put(14, "\u03C3"); // sigma
            put(30, "\u03C4"); // tau
            put(37, "\u03C5"); // upsilon
            put(11, "\u03C6"); // phi
            put(47, "\u03C7"); // chi
            put(61, "\u03C8"); // psi
            put(58, "\u03C9"); // omega

        }};

        prefix_43 = new HashMap<Integer, String>() {{

            put(63, "\u25CF"); // circle
            put(60, "SQUARE_TRIANGLE_INDICATOR");

        }};

        prefix_43_60 = new HashMap<Integer, String>() {{

            put(25, "\u25A0"); // square
            put(9, "\u25B2"); // triangle

        }};

        prefix_51 = new HashMap<Integer, String>() {{

            put(21, "\u2192"); // right arrow
            put(42, "\u2190"); // left arrow
            put(41, "\u2193"); // down arrow
            put(44, "\u2191"); // up arrow
            put(14, "\u2197"); // up right arrow
            put(35, "\u2198"); // down right arrow
            put(28, "\u2199"); // down left arrow
            put(49, "\u2196"); // up left arrow

        }};

        // not here - feet
        prefix_54 = new HashMap<Integer, String>() {{

            put(54, "\u2033"); // inches

        }};

        prefix_56 = new HashMap<Integer, String>() {{

            put(35, "\u007B"); // opening curly bracket
            put(28, "\u007D"); // curly curly bracket
            put(50, "\u2022"); // bullet point
            put(57, "\u0023"); // #
            put(33, "\u2216"); // backward slash
            put(12, "\u2215"); // forward slash
            put(38, "\u00AB"); // opening guillemet
            put(52, "\u00BB"); // closing guillemet

            put(6, "<us>"); // UNDERLINE_SYMBOL
            put(2, "<uw>"); // UNDERLINE_WORD
            put(54, "<u>"); // UNDERLINE_PASSAGE
            put(4, "</u>"); // UNDERLINE_TERMINATOR

        }};

        prefix_60 = new HashMap<Integer, String>() {{

            put(35, "\u266D"); // flat
            put(33, "\u266E"); // natural
            put(41, "\u266F"); // sharp

        }};

        numberMode = new HashMap<Integer, String>() {{

            put(35, "G1");
            put(33, "NS");

        }};

        letters = new HashMap<Integer, String>() {{

            put(1, "\u0061"); // a
            put(3, "\u0062"); // b
            put(9, "\u0063"); // c
            put(25, "\u0064"); // d
            put(17, "\u0065"); // e
            put(11, "\u0066"); // f
            put(27, "\u0067"); // g
            put(19, "\u0068"); // h
            put(10, "\u0069"); // i
            put(26, "\u006A"); // j
            put(5, "\u006B"); // k
            put(7, "\u006C"); // l
            put(13, "\u006D"); // m
            put(29, "\u006E"); // n
            put(21, "\u006F"); // o
            put(15, "\u0070"); // p
            put(31, "\u0071"); // q
            put(23, "\u0072"); // r
            put(14, "\u0073"); // s
            put(30, "\u0074"); // t
            put(37, "\u0075"); // u
            put(39, "\u0076"); // v
            put(58, "\u0077"); // w
            put(45, "\u0078"); // x
            put(61, "\u0079"); // y
            put(53, "\u007A"); // z

        }};

        numbers = new HashMap<Integer, String>() {{

            put(26, "\u0030"); // 0
            put(1, "\u0031"); // 1
            put(3, "\u0032"); // 2
            put(9, "\u0033"); // 3
            put(25, "\u0034"); // 4
            put(17, "\u0035"); // 5
            put(11, "\u0036"); // 6
            put(27, "\u0037"); // 7
            put(19, "\u0038"); // 8
            put(10, "\u0039"); // 9

        }};

        punctuation = new HashMap<Integer, String>() {{

            put(4, "\u0027"); // ' (apostrophe)
            put(2, "\u002C"); // ,
            put(36, "\u002D"); // hyphen
            put(50, "\u002E"); // .
            put(22, "\u0021"); // !
            put(6, "\u003B"); // ;
            put(18, "\u003A"); // :
            put(38, "QUESTION_MARK_OPENING_NON_SPECIFIC_QUOTATION"); // QUESTION_MARK_OPENING_NON_SPECIFIC_QUOTATION
            put(52, "'"); // // CLOSING_NON_SPECIFIC_QUOTATION
            put(41, "\u221A"); // open square root
            put(44, "CLOSE_SQUARE_ROOT"); // CLOSE_SQUARE_ROOT
            put(34, "212"); // SUBSCRIPT
            put(20, "213"); // SUPERSCRIPT
            put(12, "\u2044"); // simple numeric fraction line
            put(55, "GFO"); // GENERAL_FRACTION_OPEN_INDICATOR
            put(62, "GFC"); // GENERAL_FRACTION_CLOSE_INDICATOR
            put(49, "ST"); // SHAPE_TERMINATOR

        }};

    }
}
